### PR TITLE
Enable another giant pile of lints

### PR DESCRIPTION
**Pull request by apoelstra**
Copy all the lints from https://github.com/rust-bitcoin/rust-bitcoin/pull/3704 and then copy all the lints from my personal "interesting lints" doc.

### CODE REVIEW

File: src/lib/dongle/message.rs [hunk 1]
```diff
@@ -250,7 +251,7 @@ impl Response for WalletPublicKey {
         // parse as a secp pubkey, and rust-bitcoin will do the right thing
         // (since it understands the bip32 spec as just using "keys" and
         // always encoding them compressedly).
-        let pk = bitcoin::secp256k1::PublicKey::from_slice(&data[1..1 + pk_len])?;
+        let pk = bitcoin::secp256k1::PublicKey::from_slice(&data[1..=1 + pk_len])?;
```
Comment (roconnor) [inline src/lib/dongle/message.rs]: Are you aware that you are changing the behaviour here to enlarge the amount of data being sliced?

Comment (apoelstra) [reply src/lib/dongle/message.rs]: Derp, no, this is not intentional! Thanks.

Comment (apoelstra) [reply src/lib/dongle/message.rs]: This was not even supposed to be committed because it is clearly wrong.

**Comment by roconnor** [comment]
I ran `listunspent` and got the following:

```
…
Unlock wallet by signing fixed message 0000990F48F5D865…0000000000000000 (Bitcoin block 332802).
thread 'main' panicked at /build/source/src/lib/dongle/mod.rs:197:9:
assertion `left != right` failed: Ledger requested user authentication but we don't know how to handle that
  left: [0, 0]
 right: [0, 0]
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
   4: icboc::dongle::Dongle::sign_message
   5: icboc3d::commands::get_wallet_key_and_nonce
   6: <icboc3d::commands::listunspent::ListUnspent as icboc3d::commands::Command>::execute
   7: icboc3d::commands::execute_from_args
   8: icboc3d::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

File: src/lib/dongle/mod.rs [hunk 2]
```diff
@@ -194,9 +194,11 @@ pub trait Dongle {
             });
         }
 
-        if rev != [0, 0] {
-            panic!("Ledger requested user authentication but we don't know how to handle that");
-        }
+        assert_ne!(
```
Comment (roconnor) [inline src/lib/dongle/mod.rs]: yeah, I think the logic ended up reversed.  Before we failed if `rev != [0, 0]` but now we are asserting that `rev != [0, 0]` is required.

Comment (apoelstra) [reply src/lib/dongle/mod.rs]: Lol, yeah. Thanks for testing! Sorry for pushing this up without trying it.

**Comment by roconnor** [comment]
I successfully ran `listunspent`

**Review by apoelstra — COMMENTED** [review]
On 8c7d78d550c8c0014aa87115b512ade6eb04aff2 successfully ran local tests

**Comment by apoelstra** [comment]
Ok, checked that `listunspent` `info` and `rescan` all continue to work as expected. Will merge.

**Pull request merged**